Extended rate-model framework

janhahne opened the pull request:
This PR extends the functionality of the rate-model framework by several options that are needed to for more complex rate models. All extensions and changes are implement such that previous simulation scripts do not need to be adjusted and all added models are actually needed in ongoing projects. The credit of this PR mostly belongs to @MSenden

Quick overview:
- the `rate_neuron_ipn` model now has a parameter `lambda` for the passive decay rate and the possibility to restrict the rate to positive values via a boolean parameter `rectify_output`.

- the `rate_neuron_ipn` class is extended such that it allows to build models with multiplicative coupling with different factors for excitatory and inhibitory coupling. The template class now has three functions: `nonlinearities_.input` (`previously operator()`), `nonlinearities_.mult_coupling_ex` and `nonlinearities_.mult_coupling_in`. So the type of multiplicative coupling can be defined in the template derived classes. `return 1;` achives nonmutliplicative coupling. In order to keep the number of models low the multiplicative coupling version of a model can be implemented in the same NEST model as the nonmultiplicative coupling version and can be switched on via the boolean `mult_coupling` (see `lin_rate.h`)

- A new template class `parrot_rate_neuron` was added. The functionality is similar to that of `parrot_neuron` but for secondary events. See the documentation in the file for details.

- the `threshold_lin_rate` model is now implemented in a more general way and allows further settings.

- three additional models (`gauss_rate`, `sigm_rate` and `sigmoid_rate`) have been added.

- a decision making example that uses some of the new functionality was added

- the rate model unit test was extended to check the `lin_rate` model with multiplicative coupling.

janhahne commented:
I suggest @ddahmen and @mschmidt87 as reviewers.

ddahmen commented on models/parrot_rate_neuron.h:
```diff
@@ -0,0 +1,287 @@
+/*
+ *  parrot_rate_neuron.h
```
I find the name 'parrot_rate_neuron' a bit misleading.
In NEST, a parrot neuron is one that exactly repeats what it gets in the input.
Here you apply a nonlinear transformation. So I would rather think of it as a rate neuron with time constant tau=0.
tau=0 is currently not possible in the implementation of rate_neuron_ipn etc.
But would it not be possible to insert some if statements in the definitions of the propagators of rate_neuron_ipn to enable tau=0? Then you could spare this new model here. Although it might be also good to have these two cases, tau>0  and tau=0, separate as in the former case the neuron has some internal dynamics and in the latter case not.
If you decide for keeping this new model, I would suggest to rename it to something like 'rate_nonlinearity'.

Also keep in mind that the behavior for linear_summation=True is different here from rate_neuron_ipn/opn because here you do not keep excitatory and inhibitory inputs separate.

janhahne replied:
You are right, that the `parrot_rate_neuron` only repeats its input in the linear version, i.e. `lin_rate_parrot`. All other versions do indeed apply the corresponding nonlinearity to the input.
I like the name as it gives a hint what the neuron is used for, which is basicly the buffering (and here also the modifcation) of its received input. Anyway if @heplesser also agrees that the name is misleading or misplaced here we could indeed change it.

You are also right that the class basicly represents and behaves like a `rate_neuron_ipn` with `tau=0`. As far as I remember the main reasons why we implemented this in a seperate class (besides the reason that the integration into `rate_neuron_ipn` would require a couple of new if-statements, which would make the code even more complex) is that it makes simulation scripts much easier to read: It is much more intuitive to think of those neurons as buffer or helper neurons than to classify them as `rate_neuron_ipn` neurons with `tau=0`. Did I get this right @MSenden?

MSenden replied:
@janhahne yes, you are right. In the model whose implementation drove my choice for rate neuron types, some neurons would apply different nonlinearities to inputs originating from different sources. To handle this the outputs of the different sources are passed through parrot neurons applying the desired nonlinearities before they reach a linear rate neuron. @ddahmen You are also right that this is equivalent as using a rate neuron with the desired nonlinearity and setting tau to zero but when building a model I was under the impression that the code is easier to follow if a source's output is passed through a helper unit explicitly dedicated to instantaneously applying a nonlinearity than having to recognize another neuron as doing precisely this based on its parameter settings.

However, there is indeed no use-case for `lin_rate_parrot` as it does not apply a nonlinearity

ddahmen replied:
I agree that a model is much easier to understand if the 'parrot_rate_neuron' is not merged with the other types of neurons. So only the naming problem remains. What do you think @heplesser ?

heplesser replied:
@janhahne @MSenden @ddahmen Thank you for discussing this point in detail! If I understand the discussion correctly, then you use the "parrots" to transform rate signals on their way from a sender neuron to a target neuron. You implement this transformation as a separate network node to avoid complexities in input handling in the target neuron. This seems sensible, but I agree with @ddahmen that the naming is unfortunate, as a parrot should indeed repeat its input perfectly. I am also a bit confused by the use of the term "buffer" in the discussion, as my understanding is that we are here talking about instantaneous transformations.

I would suggest `rate_transformer_<func>` as name, where `<func>` describes the transformation function applied. If the linear version does nothing, I would drop it, but otherwise I would agree to provide all possible transformeres if adding them only requires an additional template instantiation.

janhahne replied:
I like your suggestion and implemented the change of the model name(s). Using the term "buffer" in the discussion was indeed unfortunate - i meant a "helper" neuron.

janhahne commented on models/sigm_rate.h:
```diff
@@ -0,0 +1,151 @@
+/* BeginDocumentation
+Name: sigm_rate - rate model with sigmoidal gain function as defined
+in [1].
+
```
Thank you for the reasonable suggestion. I changed the name with commit 3e1c816

janhahne commented on models/threshold_lin_rate.h:
```diff
@@ -37,13 +42,11 @@ Name: threshold_lin_rate - rate model with threshold-linear gain function
- with Heaviside function H.
+ threshold_lin_rate is an implementation of a nonlinear rate model with input
+ function input(h) = min( max( g * ( h - theta ), 0 ), alpha ). Here H denotes
+ the Heaviside function. The boolean parameter linear_summation determines
```
Corrected with 6ecc124

janhahne commented on models/rate_neuron_opn.h:
```diff
@@ -177,8 +181,15 @@ class rate_neuron_opn : public Archiving_Node
+    // by RateConnectionInstantaneous from excitatory neurons
+    std::vector< double >
+      instant_rates_in_;                  //!< buffer for rate vector received
     std::vector< double > instant_rates_; //!< buffer for rate vector received
```
Oh, good catch! Corrected with 6ecc124

janhahne commented on models/lin_rate.h:
```diff
@@ -78,37 +90,80 @@ Author: David Dahmen, Jan Hahne, Jannis Schuecker
+    , g_in_( 1.0 )
+    , theta_ex_( 1.0 )
+    , theta_in_( 1.0 )
   {
```
This seems more "natural" for me as well. Changed with 6ecc124

janhahne commented on models/gauss_rate.h:
```diff
@@ -0,0 +1,147 @@
+Description:
+
+ gauss_rate is an implementation of a nonlinear rate model with input
+ function input(h) = g * exp( ( x - mu )^2 / ( 2 * sigma^2 ) ).
```
Corrected with 6ecc124

janhahne commented on models/lin_rate.h:
```diff
@@ -78,37 +90,80 @@ Author: David Dahmen, Jan Hahne, Jannis Schuecker
+  if ( mult_coupling_ )
+  {
+    return g_ex_ * ( theta_ex_ - rate );
+  }
```
@MSenden I would be interested in your opinion on this. Any strong feelings for/against a change?

MSenden replied:
@janhahne I have no strong feelings about this but agree that having + signs everywhere is more straightforward for users.

janhahne replied:
@ddahmen After some literature research we decided to keep it as is: Grossberg uses this notation for the coupling functions consistent in all of his publications.

ddahmen replied:
OK

janhahne commented on models/gauss_rate.cpp:
```diff
@@ -0,0 +1,57 @@
+  insert_( names::rate, &nest::gauss_rate_ipn::get_rate_ );
+  insert_( names::noise, &nest::gauss_rate_ipn::get_noise_ );
+}
+
```
So far we only created parrot versions for the `lin_rate` and the `sigmoid_rate_gg_1998` (former `sigm_rate`). Our idea was to keep the zoo of different rate models as small as possible and to create the parrot versions only for models where they are actually needed in ongoing projects.

On the other hand one could of course argue that it would be better to include all of these models in the next release version, such that different networks can be simulated with the release version without touching the code.

What would you prefer @ddahmen @heplesser?

ddahmen replied:
Since, in this case, adding the neuron model only amounts to adding a few lines of code in existing files, I would prefer to include them in the first place. 
What do you think @heplesser ?

janhahne commented on models/rate_neuron_ipn_impl.h:
```diff
@@ -247,21 +268,38 @@ nest::rate_neuron_ipn< TGainfunction >::update_( Time const& origin,
       {
-        S_.rate_ += V_.P2_ * gain_( B_.delayed_rates_.get_value_wfr_update(
-                                      lag ) + B_.instant_rates_[ lag ] );
+        S_.rate_ += V_.P2_
```
@ddahmen That is a really good catch! Of course you are right about this - the new version as is would be inconsistent to NEST 2.14 for `linear_summation = True` with a nonlinear model with both inhibitory and excitatory inputs and it even disables the use of a "joint linear summation".

@jschuecker @MSenden We will need to think about how to resolve this in more detail.

janhahne replied:
@ddahmen Mario and I resolved this issue by moving the check for multiplicative coupling to the base class in ff74e56. Now everything should be consistent with NEST 2.14 again.

The diff of ff74e56 might look a bit confusing, since I rearranged some parts of the code to avoid code duplication (it is easier to follow the changes by looking at the new file).

janhahne commented:
@ddahmen Thank you for the careful review. I resolved some of your requests with the two new commits. Regarding some other points I replied to your comments and wait for further input from you, the second reviewer @heplesser or @MSenden.

I will think about how to resolve the inconsistency/loss of functionality due to the input splitting/multiplicative coupling.

janhahne commented on models/rate_neuron_ipn_impl.h:
```diff
@@ -274,14 +312,32 @@ nest::rate_neuron_ipn< TGainfunction >::update_( Time const& origin,
+          * nonlinearities_.mult_coupling_in( new_rates[ lag ] )
+          * nonlinearities_.input( B_.delayed_rates_in_.get_value( lag )
+              + B_.instant_rates_in_[ lag ] );
       }
```
should be resolved with ff74e56

ddahmen commented on models/rate_neuron_opn_impl.h:
```diff
@@ -257,13 +261,25 @@ nest::rate_neuron_opn< TGainfunction >::update_( Time const& origin,
+          * nonlinearities_.mult_coupling_in( new_rates[ lag ] )
+          * nonlinearities_.input( B_.delayed_rates_in_.get_value_wfr_update(
+                                     lag ) + B_.instant_rates_in_[ lag ] );
       }
```
Nice fix. This makes the code consistent with the old behavior. The only issue I still have is the following:
If I thought of the simplest nonlinear rate model with multiplicative coupling, I would write down a term like

f(rate)*g(input)

where f = mult_coupling_ex = mult_coupling_in and input = exc_input + inh_input.
This model is not captured by your code which is OK. But I think that the documentation of the models should clarify this. In particular it should become clear to the user that in case of multiplicative coupling and mult_coupling_ex = mult_coupling_in linear summation still does not sum up excitatory and inhibitory inputs linearly.
This would also maybe suggest to rename the boolean linear_summation to make it less confusing.
What do you think?
Sorry for being so picky...

janhahne replied:
@ddahmen You are right that this simplest multiplicative coupling can only be achieved with a trick (connect all incoming inhibitory connections with a `lin_rate_parrot` neuron with the desired weight and then connect this helper neuron with the target neuron with weight 1.0 - then the mult_coupling_ex is the joint factor for the multiplicative coupling).
We can include this in the documentation somewhere - however at the present time I don't know where to put it: The only model that has multiplicative coupling implemented is the lin_rate neuron, for which g(input_exc) + g(input_inh) = g(input_exc + input_inh) holds anyway.

I would also suggest to keep the name `linear_summation`, for consistency to 2.14 and since - given two different multiplicative coupling factors for exh and inh coupling - it should be clear that the linear summation can only happen within the two subsets (or is there some way that I do not think of at the moment?)

ddahmen replied:
You are right that the naming better be consistent with NEST 2.14. I also agree that for lin_rate the above mentioned problem is irrelevant. The only problem could be that in the future somebody will create another neuron model from the templates and does not keep the above issue in mind. One could help future developers by adding the following sentence to the documentation of rate_neuron_opn_impl and rate_neuron_ipn_impl:

"The boolean parameter linear_summation
 determines whether the input function is applied to the summed up incoming
 connections (= True, default value) or to each input individually (= False). In case of multiplicative coupling the nonlinearity is applied separately to the summed excitatory and inhibitory inputs if linear_summation=True."

janhahne replied:
I added the suggested additional comment

heplesser commented on models/sigmoid_rate.h:
```diff
@@ -0,0 +1,147 @@
+ function input(h) = g / ( 1. + exp( -beta * ( h - theta ) ) ).
+ The boolean parameter linear_summation determines whether the input
+ function is applied to the summed up incoming connections
+ (= True, default value) or to each input individually (= False).
```
See comment on documentation above.

heplesser commented on pynest/examples/rate_neuron_DM.py:
```diff
@@ -0,0 +1,166 @@
+# -*- coding: utf-8 -*-
```
I would go for lowercase `_dm` in the file name.

janhahne replied:
I agree, looks better with lowercase

janhahne commented on models/parrot_rate_neuron.h:
```diff
@@ -0,0 +1,287 @@
+#include "recordables_map.h"
+#include "universal_data_logger.h"
+
+#include <string>
```
Done

janhahne commented on models/parrot_rate_neuron.h:
```diff
@@ -0,0 +1,287 @@
+Sends: InstantaneousRateConnectionEvent, DelayedRateConnectionEvent
+
+Parameters:
+No parameters to be set in the status dictionary.
```
Yes, good catch. I updated the documentation.

janhahne commented on models/parrot_rate_neuron.h:
```diff
@@ -0,0 +1,287 @@
+Author: Mario Senden, Jan Hahne, Jannis Schuecker
+FirstVersion: November 2017
+*/
+template < class TNonlinearities >
```
`class` seems to be used in many templates at the moment (i looked for example in rate_neuron_ipn/opn and the binary neuron). Maybe this can rather be fixed in an separate PR for all NEST files?

janhahne commented on models/parrot_rate_neuron_impl.h:
```diff
@@ -0,0 +1,285 @@
+    }
+  }
+
+  if ( not called_from_wfr_update )
```
You probably missed the end of the for-loop in between ;)

janhahne commented on models/lin_rate.h:
```diff
@@ -51,9 +56,16 @@ Name: lin_rate - Linear rate model
  mean                double - Mean of Gaussian white noise.
  std                 double - Standard deviation of Gaussian white noise.
- g                   double - Gain parameter
+ g                   double - Linear factor in input function.
```
Reasonable question - i changed it back to avoid confusion

janhahne commented on models/lin_rate.h:
```diff
@@ -37,9 +38,13 @@ Name: lin_rate - Linear rate model
+ mult_coupling_ex(rate) = g_ex_ * ( theta_ex_ - rate ) and
+ mult_coupling_in(rate) = g_in_ * ( theta_in_ + rate ).
+ The multiplicative coupling can be switched on and off via
+ the boolean parameter mult_coupling (default=false).
```
I tried to improve the documentation, please tell me if the new version gives you all the information you need

janhahne commented on models/gauss_rate.h:
```diff
@@ -0,0 +1,147 @@
+ function input(h) = g * exp( -( x - mu )^2 / ( 2 * sigma^2 ) ).
+ The boolean parameter linear_summation determines whether the input
+ function is applied to the summed up incoming connections
+ (= True, default value) or to each input individually (= False).
```
Agreed. Applied this for all models

janhahne commented on models/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ set( models_sources
     dc_generator.h dc_generator.cpp
     diffusion_connection.h
     gamma_sup_generator.h gamma_sup_generator.cpp
+    gauss_rate.h gauss_rate.cpp
```
Good catch!

janhahne commented:
@heplesser Thank you for the careful review! I addressed all your comments, please have another look. 

The inconsistency was removed with one of the commits following on Davids suggestion: Now the behavior is completely consistent with NEST 2.14.0 (in case of non-multiplicative coupling - which is the only coupling available in 2.14.0 - the exc+inh input is summed up before the nonlinearity is applied).

heplesser commented:
@janhahne @MSenden @jschuecker @ddahmen Thanks a lot for your efforts! This looks fine now and I will merge it right away.
